Softeq.NetKit.Chat.Domain.Services/DomainServices/DirectChannelService.cs

```diff
@@ -71,5 +71,127 @@ public async Task<DirectChannelResponse> GetDirectChannelById(Guid id)
+        }
+
+        public async Task<DirectMessageResponse> DeleteMessageAsync(Guid id, string saasUserId)
+        {
+            var directMessage = await UnitOfWork.DirectMessagesRepository.GetMessageByIdAsync(id);
+            if (directMessage == null)
+            {
+                throw new NetKitChatNotFoundException($"Unable to get direct message. Message with {nameof(id)}:{id} is not found.");
+            }
+            var owner = await UnitOfWork.MemberRepository.GetMemberBySaasUserIdAsync(saasUserId);
+            if (owner == null)
+            {
+                throw new NetKitChatNotFoundException($"Unable to get member {nameof(saasUserId)}:{saasUserId} is not found.");
+            }
+
+            await UnitOfWork.DirectMessagesRepository.DeleteMessageAsync(id);
+
+            return DomainModelsMapper.MapToDirectMessageResponse(directMessage, owner);
+        }
+
+        public async Task<DirectMessageResponse> UpdateMessageAsync(UpdateDirectMessageRequest request)
+        {
+            var directMessage = await UnitOfWork.DirectMessagesRepository.GetMessageByIdAsync(request.MessageId);
+            if (directMessage == null)
+            {
+                throw new NetKitChatNotFoundException($"Unable to get direct message. Message with {nameof(request.MessageId)}:{request.MessageId} is not found.");
+            }
+            var owner = await UnitOfWork.MemberRepository.GetMemberBySaasUserIdAsync(request.SaasUserId);
+            if (owner == null)
+            {
+                throw new NetKitChatNotFoundException($"Unable to get member. Member { nameof(directMessage.OwnerId) }:{ directMessage.OwnerId} is not found.");
+            }
+
+            var message = new DirectMessage
+            {
+                Body = request.Body,
+                Created = directMessage.Created,
+                DirectChannelId = directMessage.DirectChannelId,
+                Id = directMessage.Id,
+                OwnerId = directMessage.OwnerId,
+                Updated = _dateTimeProvider.GetUtcNow()
+            };
+
+            await UnitOfWork.DirectMessagesRepository.UpdateMessageAsync(message);
+
+            return DomainModelsMapper.MapToDirectMessageResponse(message, owner);
+        }
+
+        public async Task<IList<DirectMessageResponse>> GetMessagesByChannelIdAsync(Guid channelId)
+        {
+            var channel = await UnitOfWork.DirectChannelRepository.GetDirectChannelById(channelId);
+            if (channel == null)
+            {
+                throw new NetKitChatNotFoundException($"Unable to add direct message. Channel { nameof(channelId) }:{ channelId} is not found.");
+            }
+
+            var messages = await UnitOfWork.DirectMessagesRepository.GetMessagesByChannelIdAsync(channelId);
+
+            var uniqueMembersId = messages.GroupBy(x => x.OwnerId).Select(x => x.FirstOrDefault()).Select(y => y.OwnerId);
```

Comment:
NullReferenceException is possible in Select

Reply:
Done